flow/project/cli/config/manipulators/composer.go

```diff
@@ -0,0 +1,210 @@
+ * limitations under the License.
+ */
+
+package manipulators
+
+import (
+	"errors"
+	"fmt"
+	"github.com/onflow/flow-cli/flow/project/cli/config"
+	"os"
+	"path/filepath"
+
+	"github.com/spf13/afero"
+)
+
+// ErrDoesNotExist is error to be returned when config file does not exists
+var ErrDoesNotExist = errors.New("project config file does not exist")
+
+// Exists checks if file exists on the specified path
+func Exists(path string) bool {
+	info, err := os.Stat(path)
+	if os.IsNotExist(err) {
+		return false
+	}
+	return !info.IsDir()
+}
+
+// Parser is interface for any configuration format parser to implement
+type Parser interface {
+	Serialize(*config.Config) ([]byte, error)
+	Deserialize([]byte) (*config.Config, error)
+	SupportsFormat(string) bool
+}
+
+type ConfigParsers []Parser
+
+func (c *ConfigParsers) FindForFormat(extension string) Parser {
+	for _, parser := range *c {
+		if parser.SupportsFormat(extension) {
+			return parser
+		}
+	}
+
+	return nil
+}
+
+type Composer struct {
+	af               *afero.Afero
+	configParsers    ConfigParsers
+	composedMultiple bool
+	composedFromFile map[string]string
+}
+
+func NewComposer(filesystem afero.Fs) *Composer {
+	af := &afero.Afero{Fs: filesystem}
+	return &Composer{
+		af:               af,
+		composedFromFile: map[string]string{},
+	}
+}
+
+func (c *Composer) AddConfigParser(format Parser) {
+	c.configParsers = append(c.configParsers, format)
+}
+
+func (c *Composer) Save(conf *config.Config, path string) error {
+	if c.composedMultiple || c.composedFromFile != nil {
+		return errors.New("Saving configuration to multiple files currently not supported")
+	}
+
+	configFormat := c.configParsers.FindForFormat(
+		filepath.Ext(path),
+	)
+
+	data, err := configFormat.Serialize(conf)
+	if err != nil {
+		return err
+	}
+
+	err = c.af.WriteFile(path, data, 0644)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Load and compose multiple configurations
+func (c *Composer) Load(paths []string) (*config.Config, error) {
+	var baseConf *config.Config
+
+	for _, path := range paths {
+		raw, err := c.loadFile(path)
+		if err != nil {
+			return nil, err
+		}
+
+		preProcessed := c.preprocess(raw)
+		configParser := c.configParsers.FindForFormat(filepath.Ext(path))
+		if configParser == nil {
+			return nil, fmt.Errorf("Parser not found for config: %s", path)
+		}
+
+		conf, err := configParser.Deserialize(preProcessed)
+		if err != nil {
+			return nil, err
+		}
+
+		// if first conf just assign as baseConf
+		if baseConf == nil {
+			baseConf = conf
+			continue
+		}
+
+		c.composeConfig(baseConf, conf)
+	}
+
+	baseConf, err := c.postprocess(baseConf)
+	if err != nil {
+		return nil, err
+	}
+
+	return baseConf, nil
+}
+
+func (c *Composer) AddAccountFromFile(path string, name string) {
+	c.composedFromFile[name] = path
+}
+
+func (c *Composer) preprocess(raw []byte) []byte {
+	preprocessor := NewPreprocessor(c)
```

Comment:
I noticed there's a hidden mutation to `c.composedFromFile` inside of `Preprocessor.Run`, called through `AddAccountFromFile`. I think we should avoid function side effects like this.

Instead of mutating `c.composedFromFile`, could the `Run` function just return a list containing the account files discovered?

```go

type accountFromFile struct {
  name string
  path string
}

allAccountFiles = make([]accountFromFile)

for _, path := range paths {
  // ...

  preProcessed, accountFiles := c.preprocess(raw)
  
  allAccountFiles = append(allAccountFiles, accountFiles...)

  // ...
}

baseConf, err := c.postprocess(baseConf, allAccountFiles)
if err != nil {
  return nil, err
}
```

Reply:
Yes, much better this way.